### PR TITLE
feat: add encryption for api keys

### PR DESCRIPTION
**Pull request by Harish-osmosys**
### Pre-requisites

- [x] I have gone through the Contributing guidelines for [Submitting a Pull Request (PR)](https://github.com/OsmosysSoftware/osmo-x/blob/main/CONTRIBUTING.md#-submitting-a-pull-request-pr) and ensured that this is not a duplicate PR.
- [x] I have performed unit testing for the new feature added or updated to ensure the new features added are working as expected.
- [ ] I have added/updated test cases to the [test suite](https://github.com/OsmosysSoftware/osmo-x/blob/main/apps/api/OsmoX.postman_collection.json) as applicable.
- [x] I have performed preliminary testing using the [test suite](https://github.com/OsmosysSoftware/osmo-x/blob/main/apps/api/OsmoX.postman_collection.json) to ensure that any existing features are not impacted and any new features are working as expected as a whole.
- [ ] I have added/updated the required [api docs](https://github.com/OsmosysSoftware/osmo-x/tree/main/apps/api/docs) as applicable.
- [ ] I have added/updated the `.env.example` file with the required values as applicable.

### PR Details

PR details have been updated as per the given format (see below)

- [x] PR title adheres to the format specified in guidelines (e.g., `feat: add admin login endpoint`)
- [x] Description has been added
- [x] Related changes have been added (optional)
- [x] Screenshots have been added (optional)
- [x] Query request and response examples have been added (as applicable, in case added or updated)
- [ ] Documentation changes have been listed (as applicable)
- [ ] Test suite/unit testing output is added (as applicable)
- [ ] Pending actions have been added (optional)
- [ ] Any other additional notes have been added (optional)

### Additional Information

- [x] Appropriate label(s) have been added (`ready for review` should be added if the PR is ready to be reviewed)
- [ ] Assignee(s) and reviewer(s) have been added (optional)

_Note: Reviewer should ensure that the checklist and description have been populated and followed correctly, and the PR should be merged only after resolving all conversations and verifying that CI checks pass._

---

**Description:**

This Pr does the following things
- Add a new graph api to generate new server api keys
- Add encryption to the api key
- Add a new column MaskedKey
- update the logic to verify the key 

**Related changes:**

- Add new migration to add a masked Key column in server-api table 
- 
**Screenshots:**
// api 

image

//db
image

**Query request and response:**

- cURL to add new server api key 
```
curl --location 'http://localhost:3000/graphql' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6IkFkbWluIiwidXNlcklkIjoxLCJyb2xlIjoxLCJpYXQiOjE3MjcxNzQ5OTMsImV4cCI6MTcyOTc2Njk5M30.a6g-OIdUcAZiyNcJ_AxRgyIWEbGnVnWRK5EVOsuJEWQ' \
--data '{"query":"mutation GenerateApiKey($applicationId: Int!) {\n  generateApiKey(applicationId: $applicationId)\n}","variables":{"applicationId":1}}'
```
**Documentation changes:**

NA
**Test suite/unit testing output:**

NA
**Pending actions:**

NA
**Additional notes:**

NA

## Release Notes

- **New Features**
  - Introduced a new method for generating random API keys for applications, enhancing user access to API functionality.
  - Added a masked API key feature for improved security in the API key management process.

- **Improvements**
  - Enhanced API key validation logic to support multiple keys for flexibility.
  - Updated password and API key handling for better security and configurability.

- **Bug Fixes**
  - Adjusted the return type of API key retrieval methods to ensure accurate data handling.

These updates aim to improve security, flexibility, and user experience within the application.

### CODE REVIEW

File: apps/api/src/common/utils/bcrypt.ts [hunk 1]
```diff
@@ -1,11 +1,35 @@
+import { ConfigService } from '@nestjs/config';
 import * as bcrypt from 'bcrypt';
 
+const configService = new ConfigService();
+
+// Define your salt rounds
+const SALT_ROUNDS = configService.get<number>('SALT_ROUNDS');
+const SECRET = configService.get<string>('API_KEY_SECRET');
```
[the remark this answers is not in the document]
Comment (Harish-osmosys) [reply apps/api/src/common/utils/bcrypt.ts:8]: as this is simple utitility class no need to create class for this

File: apps/api/src/common/guards/api-key/api-key.guard.ts [hunk 2]
```diff
@@ -102,22 +95,18 @@
       throw new BadRequestException(`Provider ${providerEntry.name} is not enabled`);
     }
 
-    // Set correct ApplicationId after verifying
-    const inputApplicationId = await this.getApplicationIdFromApiKey(apiKeyToken);
-    this.logger.debug(
-      `Fetched ApplicationId from DB using APIKeyToken: ${JSON.stringify(inputApplicationId)}`,
+    const apiKeys = await this.serverApiKeysService.findByRelatedApplicationId(
+      providerEntry.applicationId,
```
[the remark this answers is not in the document]
Comment (Harish-osmosys) [reply apps/api/src/common/guards/api-key/api-key.guard.ts:99]: added this change in the latest commit

File: apps/api/src/modules/server-api-keys/server-api-keys.service.ts [hunk 3]
```diff
@@ -19,8 +21,29 @@
...
+    return this.serverApiKeyRepository.find({
+      where: {
+        applicationId,
+        status: Status.ACTIVE,
+      },
+    });
+  }
+
+  async generateApiKey(applicationId: number): Promise<string> {
+    const originalApiKey = await bcrypt.genSalt();
```
[the remark this answers is not in the document]
Comment (Harish-osmosys) [reply apps/api/src/modules/server-api-keys/server-api-keys.service.ts]: Updated to use crypto to generate randome key

File: apps/api/src/modules/server-api-keys/server-api-keys.service.ts [hunk 4]
```diff
@@ -19,8 +21,29 @@ export class ServerApiKeysService extends CoreService<ServerApiKey> {
...
+    const encryptedApiKey = await encryptApiKey(originalApiKey);
```
[the remark this answers is not in the document]
Comment (Harish-osmosys) [reply apps/api/src/modules/server-api-keys/server-api-keys.service.ts]: Update function name

**Review by xixas — CHANGES_REQUESTED** [review]
Check comments from code rabbit too.

**Review by xixas — APPROVED** [review]

**Pull request merged**